app/src/main/java/xyz/whereuat/whereuat/LoginActivity.java

```diff
@@ -196,7 +197,15 @@ private boolean isValidPhoneForm(String phone_number) {
      * after generating the GCM token.
      */
     private class TokenBroadcastReceiver extends BroadcastReceiver {
+        private HashSet<Integer> received_ids = new HashSet<>();
+
         public void onReceive(Context context, Intent intent) {
+            // Check the unique broadcast ID. If it is a repeat, ignore the broadcast.
+            Integer broadcast_id = intent.getIntExtra(Constants.TOKEN_BROADCAST_ID_EXTRA, 0);
```

Comment:
make the default -1 so we're positive something messed up when we do the check on line 205. `Random.nextInt(n)` is from [0, n)